src/gluonnlp/models/electra.py

```diff
@@ -531,72 +476,49 @@ class ElectraGenerator(HybridBlock):
-                                              weight_initializer=weight_initializer,
-                                              bias_initializer=bias_initializer,
-                                              prefix='proj_'))
-                self.mlm_decoder.add(get_activation(self.backbone_model.activation))
-                self.mlm_decoder.add(nn.LayerNorm(epsilon=self.backbone_model.layer_norm_eps,
-                                                  prefix='ln_'))
-                # only load the dense weights with a re-initialized bias
-                # parameters are stored in 'word_embed_bias' which is
-                # not used in original embedding
-                self.mlm_decoder.add(
-                    nn.Dense(
-                        units=self.backbone_model.vocab_size,
-                        flatten=False,
-                        params=self.backbone_model.word_embed.collect_params('.*weight'),
-                        bias_initializer=bias_initializer,
-                        prefix='score_'))
-            self.mlm_decoder.hybridize()
+        super().__init__()
+        self.backbone_model = ElectraModel.from_cfg(backbone_cfg)
+        if weight_initializer is None:
+            weight_initializer = self.backbone_model.weight_initializer
+        if bias_initializer is None:
+            bias_initializer = self.backbone_model.bias_initializer
+        self.mlm_decoder = nn.HybridSequential()
+        # Extra non-linear layer
+        self.mlm_decoder.add(nn.Dense(units=self.backbone_model.embed_size,
+                                      flatten=False,
+                                      weight_initializer=weight_initializer,
+                                      bias_initializer=bias_initializer))
+        self.mlm_decoder.add(get_activation(self.backbone_model.activation))
+        self.mlm_decoder.add(nn.LayerNorm(epsilon=self.backbone_model.layer_norm_eps))
+        # only load the dense weights with a re-initialized bias
+        # parameters are stored in 'word_embed_bias' which is
+        # not used in original embedding
+        self.mlm_decoder.add(
+            nn.Dense(
+                units=self.backbone_model.vocab_size,
+                flatten=False,
+                bias_initializer=bias_initializer))
+        self.mlm_decoder[-1].weight = self.backbone_model.word_embed.weight
+        self.mlm_decoder.hybridize()
+
+    def tie_embeddings(self, word_embed_params=None, token_type_embed_params=None,
+                       token_pos_embed_params=None, embed_layer_norm_params=None):
+        self.backbone_model.word_embed.share_parameters(word_embed_params)
+        self.mlm_decoder[-1].share_parameters(word_embed_params)
+        self.backbone_model.token_type_embed.share_parameters(token_type_embed_params)
+        self.backbone_model.token_pos_embed.share_parameters(token_pos_embed_params)
+        self.backbone_model.embed_layer_norm.share_parameters(embed_layer_norm_params)
```

Comment:
From the above code, we can share weights in two ways: 1. weight = weight 2.share_parameters, is that correct? Do we need consistency on this part?

Reply:
Yes, both is supported. The `1. weight = weight` way works well if you only need to replace a single parameter and you know it's location. The second way is more general and doesn't require users to specify the position of the weight (but just pass a dictionary containing the weights with proper name).